content/subtleties.md

```diff
@@ -14,37 +14,39 @@ Also see Pierre Weis's [Frequently asked Questions about Caml](http://caml.inria
+  
+  In OCaml and other ML-style languages that support mutability, only _values_ can have polymorphic types, hence the name -- "value restriction". The underlying reason, is that if an expression is mutable (an since we don't know its type, we must assume that it is mutable), then it could be assigned multiple times, each time with a value of a different type, thus we can't guarantee that the type of a value would be preserved. That will eventually break the type soundness. 
+  
+  Originally, values were defined as syntatic constants. I.e., something that is a static constant on the source level, is considered a value, e.g., `1`, `"hello"`, `None`, `fun x -> x`. However, in OCaml this restriction is relaxed, and besides syntactic constants OCaml allows some other classes of expressions to be generalized (i.e., to have the polymoprhic type). First of all, OCaml allows polymorphism for _non-expansive_ expressions, where an expression is non-expansive if it doesn't have any observable side-effects during the evaluation. That basically extends the notion of constant from purely syntactic definition (something that looks like a constant, is a constant) to semantic definition (something that acts like a constant, is a constant). For example, this restiction allows the following expressions to have polymorphic type: `lazy None`, ` let _ = ref None in 1`, `assert false`. But that's not all, OCaml is able to prove that it is safe to generalize certain type variables, by performing their variance analysis. It was [observed]((https://caml.inria.fr/pub/papers/garrigue-value_restriction-fiwflp04.pdf)) that there is a nice interaction between subtyping, variance, and value restriction. If a value occurs only in covariant positions, then it can be safely generalized and given the polymorphic type. With this rule, it is possible to give the polymorphic type to the following expressions: `fun () -> []`, `fun () -> ref None`. 
+  
+  The value restriction is a pessimistic assumption that a compiler must make in the face of the possible mutability, and is a price that must be payed even if mutability is not used. Weak type variables alleviate the problem a little, by allowing us to omit type annotations for values that don't have concrete or polymorphic type. However, since types are analyzed on the compilation unit level (the file level), weak type variables can't escape the scope of a compilation unit (the scope of a file). So it is not possible to have a weak type variable in the module interface (e.g., in the `mli` file). Compiler will reject all toplevel definitions that have weak types, unless they are either hidden from the module interface (i.e., omitted in the interface - e.g., when the mli file is empty) or given a concrete type. 
```

Comment:
I'd paragraph break just before `However`, and again, please put in line breaks into the markdown source, it will help editing. Also, `Compiler` should be `The compiler`.

Reply:
If you edit in github (which may not be what you want), it will match indents with previous lines.